libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt

```diff
@@ -1505,6 +1527,67 @@ internal class MapRouteLine(
                 else -> y2
             }
         }
+
+        fun getRouteLineTrafficExpressionData(route: DirectionsRoute):
+            List<RouteLineTrafficExpressionData> {
+                val distances = route.legs()?.mapNotNull {
+                    it.annotation()?.distance()
+                }?.flatten()
+                ifNonNull(distances) { distanceList ->
+                    val roadClassMap = mutableMapOf<Int, String>()
+                    route.legs()?.asSequence()
+                        ?.map { it.steps() }
+                        ?.filterNotNull()
+                        ?.flatten()
+                        ?.mapNotNull { it.intersections() }
+                        ?.flatten()
+                        ?.filter {
+                            it.geometryIndex() != null && it.mapboxStreetsV8()?.roadClass() != null
+                        }?.toList()?.forEach {
+                            roadClassMap[it.geometryIndex()!!] =
+                                it.mapboxStreetsV8()!!.roadClass()!!
+                        }
+
+                    return route.legs()?.mapNotNull {
+                        it.annotation()?.congestion()
+                    }?.flatten()?.mapIndexed { index, congestion ->
+                        val distanceOffset = distanceList.subList(0, index).sum()
```

Comment:
While trying to draw a route from Wrocław to Paris, it takes ~25 seconds for it to show up while the main thread is frozen. Profiling shows that this line is consuming ~61% of the main thread's time.
![Screenshot from 2020-10-26 12-03-25](https://user-images.githubusercontent.com/16925074/97165336-d5335d80-1783-11eb-9463-a85305d63c6b.png)

Would you be able to give all of the calculations from this PR a pass and optimize them?

Reply:
Good catch. One bug I found was that the calculation of the traffic congestion was being done twice. So eliminating that improved things quite a bit.  I also removed the numerous calls to sum the distance of the route sections and instead keep a running total of the distance as the code iterates through the route sections.  In addition the code looks for duplicate congestion/road class entries to create a simpler expression for the gradient line. So for example if the first route section has a congestion level = low and the second route section has the same congestion level and road class I omit the second item in the gradient expression since it wouldn't change the appearance of the gradient line.  When I create a route from San Francisco to the middle of the U.S. It is taking less than 25ms on my phone.